src/test/scala/mesosphere/mesos/TaskBuilderTest.scala

```diff
@@ -82,6 +82,83 @@ class TaskBuilderTest extends MarathonSpec {
       assert("*" == r.getRole)
     }
 
+    assert(!taskInfo.hasLabels)
+
+    // TODO test for resources etc.
+  }
+
+  test("BuildIfMatchesWithLabels") {
+    val offer = makeBasicOffer(cpus = 1.0, mem = 128.0, disk = 2000.0, beginPort = 31000, endPort = 32000)
+      .addResources(ScalarResource("cpus", 1))
+      .addResources(ScalarResource("mem", 128))
+      .addResources(ScalarResource("disk", 2000))
+      .build
+
+    val labels = Map("foo" -> "bar", "test" -> "test")
+
+    val task: Option[(TaskInfo, Seq[Long])] = buildIfMatches(
+      offer,
+      AppDefinition(
+        id = "/product/frontend".toPath,
+        cmd = Some("foo"),
+        cpus = 1,
+        mem = 64,
+        disk = 1,
+        executor = "//cmd",
+        ports = Seq(8080, 8081),
+        labels = labels
+      )
+    )
+
+    assert(task.isDefined)
+
+    val (taskInfo, taskPorts) = task.get
+    val range = taskInfo.getResourcesList.asScala
+      .find(r => r.getName == Resource.PORTS)
+      .map(r => r.getRanges.getRange(0))
+    assert(range.isDefined)
+    // The taskName is the elements of the path, reversed, and joined by dots
+    assert("frontend.product" == taskInfo.getName)
+    assert(2 == taskPorts.size)
+    assert(taskPorts(0) == range.get.getBegin.toInt)
+    assert(taskPorts(1) == range.get.getEnd.toInt)
+
+    assert(!taskInfo.hasExecutor)
+    assert(taskInfo.hasCommand)
+    val cmd = taskInfo.getCommand
+    assert(cmd.getShell)
+    assert(cmd.hasValue)
+    assert(cmd.getArgumentsList.asScala.isEmpty)
+    assert(cmd.getValue == "foo")
+
+    assert(cmd.hasEnvironment)
+    val envVars = cmd.getEnvironment.getVariablesList.asScala
+    assert(envVars.exists(v => v.getName == "HOST" && v.getValue == offer.getHostname))
+    assert(envVars.exists(v => v.getName == "PORT0" && v.getValue.nonEmpty))
+    assert(envVars.exists(v => v.getName == "PORT1" && v.getValue.nonEmpty))
+    assert(envVars.exists(v => v.getName == "PORT_8080" && v.getValue.nonEmpty))
+    assert(envVars.exists(v => v.getName == "PORT_8081" && v.getValue.nonEmpty))
+
+    val exposesFirstPort =
+      envVars.find(v => v.getName == "PORT0").get.getValue == envVars.find(v => v.getName == "PORT_8080").get.getValue
+    assert(exposesFirstPort)
+    val exposesSecondPort =
+      envVars.find(v => v.getName == "PORT1").get.getValue == envVars.find(v => v.getName == "PORT_8081").get.getValue
+    assert(exposesSecondPort)
+
+    for (r <- taskInfo.getResourcesList.asScala) {
+      assert("*" == r.getRole)
+    }
+
+    val expectedLabels = Labels.newBuilder.addAllLabels(
+      labels.map {
+        case (key, value) =>
+          Label.newBuilder.setKey(key).setValue(value).build()
+      }.asJava
+    ).build()
+    assert(taskInfo.hasLabels)
+    assert(taskInfo.getLabels == expectedLabels)
+
     // TODO test for resources etc.
```

Comment:
You copied even the // TODO ;)

I wonder if there is an alternative to copying all this asserts and code. Maybe extract it to a method? You copied a test which was already pretty long.